linera-core/src/join_set_ext.rs

```diff
@@ -5,113 +5,136 @@
+            let future = async move {
+                let _ = send_output.send(future.await);
+                let _ = send_done.send(());
+            };
+            self.0.push(recv_done);
+            wasm_bindgen_futures::spawn_local(
+                future::Abortable::new(future, abort_registration).map(drop),
+            );
+
+            TaskHandle {
+                output_receiver: recv_output,
+                abort_handle,
+            }
+        }
+
+        async fn await_all_tasks(&mut self) {
+            stream::iter(&mut self.0)
+                .then(|x| x)
+                .map(drop)
+                .collect()
+                .await
+        }
+
+        fn reap_finished_tasks(&mut self) {
+            for mut done in self.0.drain(..) {
+                while done.try_recv() == Ok(None) {}
+            }
         }
     }
+}
 
-    async fn await_all_tasks(&mut self) {
-        while self.join_next().await.is_some() {}
+#[cfg(not(web))]
+mod implementation {
+    pub use tokio::task::AbortHandle;
+
+    use super::*;
+
+    pub type JoinSet = tokio::task::JoinSet<()>;
+
+    /// An extension trait for the [`JoinSet`] type.
+    #[trait_variant::make(Send)]
+    pub trait JoinSetExt: Sized {
+        /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn`].
+        ///
+        /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+        /// [`AbortHandle`] to cancel execution of the task.
+        fn spawn_task<F: Future<Output: Send> + Send + 'static>(
+            &mut self,
+            future: F,
+        ) -> TaskHandle<F::Output>;
+
+        /// Awaits all tasks spawned in this [`JoinSet`].
+        async fn await_all_tasks(&mut self);
+
+        /// Reaps tasks that have finished.
+        fn reap_finished_tasks(&mut self);
     }
 
-    fn reap_finished_tasks(&mut self) {
-        while self.try_join_next().is_some() {}
+    impl JoinSetExt for JoinSet {
+        fn spawn_task<F>(&mut self, future: F) -> TaskHandle<F::Output>
+        where
+            F: Future + Send + 'static,
+            F::Output: Send,
+        {
+            let (output_sender, output_receiver) = oneshot::channel();
+
+            let abort_handle = self.spawn(async move {
+                let _ = output_sender.send(future.await);
+            });
+
+            TaskHandle {
+                output_receiver,
+                abort_handle,
+            }
+        }
+
+        async fn await_all_tasks(&mut self) {
+            while self.join_next().await.is_some() {}
+        }
+
+        fn reap_finished_tasks(&mut self) {
+            while self.try_join_next().is_some() {}
+        }
     }
 }
 
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::FutureExt as _;
+pub use implementation::*;
+
 /// A handle to a task spawned with [`JoinSetExt`].
 ///
 /// Dropping a handle aborts its respective task.
```

Comment:
Dropping a [tokio::task::AbortHandle](https://docs.rs/tokio/latest/tokio/task/struct.AbortHandle.html) does not abort the task. Probably the same is true for [futures::future::AbortHandle](https://docs.rs/futures/latest/futures/future/struct.AbortHandle.html), but there it's not documented.

Reply:
Oh yes I thought this was fishy but then I forgot to change it 😅 Thanks!